### PR TITLE
Fix stencil ref typing again

### LINKED ISSUE #158: All games are no longer booting after April 2 build (190cef9) on Vulkan/Linux with Proton/Wine
### Validation

- [X] I've read the [FAQ](https://github.com/xenia-project/xenia/wiki/FAQ).
- [X] The Xenia build used is from the master branch (not MLBS/AlexVS/Canary/pull requests, etc.)
- [X] This issue isn't for tech support (help with Xenia).
- [X] If this issue occurs in a specific game, I've done analysis to locate the faulty subsystem of the emulator and a potential reason in it.
- [X] I've checked if this issue hasn't already been reported.
- [X] My device meets the minimum requirements: https://github.com/xenia-project/xenia/wiki/Quickstart#system-requirements
- [X] (If building) I have read the building doc: https://github.com/xenia-project/xenia/blob/master/docs/building.md

### Describe what's going wrong

I don't know if you're accepting reports of Xenia running under Proton/Wine on Linux at all, but since Xenia Canary works really well with Lost Odyssey on earlier builds as of at least March 13 `canary_experimental@8678becda`, I thought that maybe there could be something useful here anyway. **If not, I apologize profoundly; please close this issue.**

Running Xenia Canary `canary_experimental@9992d20` (and earlier builds), and mainline master, I get to the black screen with "Loading", but 100% of the time are then presented with an crash error. This is on Steam Deck hardware with the experimental Vulkan renderer.

Now, this error seems to originate from Wine (`../src-wine/dlls/winevulkan/loader_thunks.c`), so I'm not sure if it belongs here. Nonetheless, maybe the difference with the March 13 build working well with the same Proton versions could tell you something, or anything at all.

### Describe what should happen

On build `canary_experimental@8678becda` from March 13, the game loads without major issues and good performance (pretty stable 30 FPS) with only minor graphical glitching (red artifacts on the right side of the screen, occasional "noise" on character textures).

This is on Steam Deck hardware running the experimental Vulkan renderer.

## On Proton 8.0-1 and Experimental

    Assertion failed!

    Program: Z:\home\deckGames\xenia_latest\xenia_canary.exe
    File: ../src-wine/dlls/winevulkan/loader_thunks.c
    Line: 2703

    Expression: !status

    Press OK to exist the program, or Cancel to start the Wine debugger.

## On Proton 7.0-6 and 6.3-8

    Unhandled Exception in Xenia
    <Nested Exception Encountered>
    Faulting thread name: GPU Commands (010000004)

### If applicable, upload a logfile and link it here

https://pastebin.com/8buXzDBj

### PR DESCRIPTION
**Pull request by yeah-its-gloria**
something something déjà-vu

Closes #158 (take that `wontfix`)

I tested using Sonic Unleashed (I don't have any other games on hand).
| Platform | GPU | Drivers | Prior issue
|---------|---|---|---|
| Windows | NVIDIA RTX 2060 | NVIDIA drivers | - |
| Linux   | AMD RX 6650 XT | Mesa/RADV | RADV crashes during vkCreateGraphicsPipeline, causes assert |
| Linux   | Intel UHD 620 | Mesa/ANV | blacked out graphics |

Fixes all of the issues mentioned above.

### CODE REVIEW

**Comment by Triang3l** [comment]
The correct fix would be to change `output_fragment_stencil_ref` itself to `int` since `gl_FragStencilRefARB` is `int` according to the [`GL_ARB_shader_stencil_export` documentation](https://registry.khronos.org/OpenGL/extensions/ARB/ARB_shader_stencil_export.txt).

**Comment by Triang3l** [comment]
The GPU part of d62fe21d47985e96d298dfa0dcce933fbc169380 from #139 (by the way, please don't place changes to completely unrelated subsystems in the same commit) needs to be reverted for `output_fragment_stencil_ref` to be `int` consistently — the incorrect type in store instructions was fixed some time after that change on the master branch, then the fix was integrated, but the original canary-specific non-GLSL-spec-adhering fix remained in the repository because it was never on the master branch.

**Comment by yeah-its-gloria** [comment]
> The correct fix would be to change `output_fragment_stencil_ref` itself to `int` since `gl_FragStencilRefARB` is `int` according to the [`GL_ARB_shader_stencil_export` documentation](https://registry.khronos.org/OpenGL/extensions/ARB/ARB_shader_stencil_export.txt).

Guess my commit message fell on deaf ears. Here goes:

> For consistency with the ref parameter to glStencilFunc, which is a GLint. [...] it makes no practical difference whether it's signed or unsigned [...]. An interesting note is that the GL specification states that the comparison made in the stencil test is unsigned.

The extension itself doesn't care about whether it's signed or not. Oh and uh, just to note, **Vulkan ain't GL**.  
And on that note:
> [...] non-GLSL-spec-adhering fix [...]

I assume you're well aware that using GLSL in SPIR-V/Vulkan has different rules than using it in GL, so I don't know why we'd reference GL's documentation about a shader extension that doesn't even directly apply to SPIR-V or Vulkan.

The [SPIR-V extension spec](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_shader_stencil_export.html) doesn't mentioned signedness:
> FragStencilRefEXT must only decorate output variable whose type is an arbitrary-sized integer type scalar.

According to the [format documentation](https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/chap41.html#formats-definition), these formats support storing stencil information:
- `VK_FORMAT_S8_UINT`
- `VK_FORMAT_D16_UNORM_S8_UINT`
- `VK_FORMAT_D24_UNORM_S8_UINT`
- `VK_FORMAT_D32_SFLOAT_S8_UINT`

Noticed the `UINT`? Beyond that, all functions in Vulkan use `uint32_t` to store stencil related data.

tl;dr I cannot fathom why you're throwing OpenGL documentation at a Vulkan renderer.

**Comment by Triang3l** [comment]
In theory any signedness should be okay, but in practice most SPIR-V shaders in the world (especially in applications GPU vendors actually consider high-priority like AAA games) are compiled with either glslang or DXC, or produced by DXVK or VKD3D (although these two are more likely to switch to safer patterns by themselves if something goes wrong, since it's much simpler for them to release an update, instead of trying to influence driver developers), and thus drivers are _production_-tested mostly on the patterns generated by them. Khronos's Vulkan Conformance Tests Suite also uses shaders written in GLSL, so cases not attainable via glslang are also unlikely to get sufficient testing from it.

There was a report from a Skyline developer that, if I recall correctly, geometry shaders didn't work on Qualcomm's proprietary driver if you didn't place system vertex outputs in a `gl_PerVertex`-like structure, which is not required in SPIR-V, but what glslang emits. In Xenia, we've been using `gl_PerVertex` from the beginning, so we didn't encounter the same issue.

And until very recently Xenia itself was completely broken on VKD3D because the combination of quotient, remainder and source addresses in its hand-emitted `udiv` DXBC instruction was not what FXC had ever been seen emitting by VKD3D developers, even though it seemed to be perfectly valid from the DXBC syntax perspective — but in VKD3D, which translates one `udiv` into two SPIR-V instructions `UDiv` and `UMod`, the source operands were read twice, but in between a source was overwritten with the result of one of the two.

The image format doesn't apply at this point in the pipeline because it's a part of the internal behavior of the stencil test, not of the interface of the fragment shader.

In this specific situation, we should recheck what glslang outputs for the intended usage of the stencil reference output, and reproduce that as closely as possible.

[Be conservative in what you do, be liberal in what you accept from others.](https://en.wikipedia.org/wiki/Robustness_principle)

**Comment by yeah-its-gloria** [comment]
Alright, did the testing you asked for.

Using DXC, `dxc -spirv shader.hlsl -E main -Fo out.spv -T ps_6_7`,
```hlsl
#pragma shader_stage(fragment)

uint main() : SV_STENCILREF {
    return uint(1);
}
```
gives us
```spv
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 9
; Schema: 0
               OpCapability Shader
               OpCapability StencilExportEXT
               OpExtension "SPV_EXT_shader_stencil_export"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main" %2
               OpExecutionMode %main OriginUpperLeft
               OpSource HLSL 670
               OpName %main "main"
               OpDecorate %2 BuiltIn FragStencilRefEXT
       %uint = OpTypeInt 32 0
     %uint_1 = OpConstant %uint 1
%_ptr_Output_uint = OpTypePointer Output %uint
       %void = OpTypeVoid
          %7 = OpTypeFunction %void
          %2 = OpVariable %_ptr_Output_uint Output
       %main = OpFunction %void None %7
          %8 = OpLabel
               OpStore %2 %uint_1
               OpReturn
               OpFunctionEnd
```
Using `glslc --target-spv=spv1.0 -c shader.glsl -o out.spv`,
```glsl
#version 460
#pragma shader_stage(fragment)
#extension GL_ARB_shader_stencil_export : enable

void main() {
    gl_FragStencilRefARB = 1;
}
```
gives us
```spv
; SPIR-V
; Version: 1.0
; Generator: Google Shaderc over Glslang; 11
; Bound: 10
; Schema: 0
               OpCapability Shader
               OpCapability StencilExportEXT
               OpExtension "SPV_EXT_shader_stencil_export"
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main" %gl_FragStencilRefARB
               OpExecutionMode %main OriginUpperLeft
               OpExecutionMode %main StencilRefReplacingEXT
               OpSource GLSL 460
               OpSourceExtension "GL_ARB_shader_stencil_export"
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %main "main"
               OpName %gl_FragStencilRefARB "gl_FragStencilRefARB"
               OpDecorate %gl_FragStencilRefARB BuiltIn FragStencilRefEXT
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
%_ptr_Output_int = OpTypePointer Output %int
%gl_FragStencilRefARB = OpVariable %_ptr_Output_int Output
      %int_1 = OpConstant %int 1
       %main = OpFunction %void None %3
          %5 = OpLabel
               OpStore %gl_FragStencilRefARB %int_1
               OpReturn
               OpFunctionEnd
```

Using `int` instead of `uint` in HLSL makes DXC pick signed integers instead.  
After some testing, I discovered that Mesa is happy with either typing as long as it stays consistent.

I'd argue that, since Vulkan uses unsigned integers basically everywhere else regarding stencil information, it'd make more sense to keep that, but despite that, I've reversed the change here and changed it to use `int` instead of `uint`.

**Review by eradzh87 — APPROVED** [review]

**Comment by eradzh87** [comment]
> something something déjà-vu
> 
> Closes #158 (take that `wontfix`)
> 
> I tested using Sonic Unleashed (I don't have any other games on hand).
> | Platform | GPU | Drivers | Prior issue
> |---------|---|---|---|
> | Windows | NVIDIA RTX 2060 | NVIDIA drivers | - |
> | Linux   | AMD RX 6650 XT | Mesa/RADV | RADV crashes during vkCreateGraphicsPipeline, causes assert |
> | Linux   | Intel UHD 620 | Mesa/ANV | blacked out graphics |
> 
> Fixes all of the issues mentioned above.
